File: daily_problems/2026/08/0819/personal_submission/cf104544a_yawn_sean.py
# Submission link: https://codeforces.com/gym/104544/submission/387533752
def main():
    n, x = MII()
    
    factors = []
    
    for i in range(1, 100000):
        if x // i < i: break
        if x % i == 0:
            factors.append(i)
            if x // i != i:
                factors.append(x // i)
    
    factors.sort()
    
    k = len(factors)
    d = {v: i for i, v in enumerate(factors)}
    
    cnt = [0] * k
    
    for v in MII():
        cnt[d[math.gcd(v, x)]] += 1
    
    gcds = [[0] * k for _ in range(k)]
    
    for i in range(k):
        for j in range(k):
            gcds[i][j] = math.gcd(factors[i], factors[j])
    
    dp = [inf] * k
    dp[0] = 0
    
    for i in range(k):
        for _ in range(cnt[i]):
            ndp = dp[:]
            flg = False
            
            for j in range(k):
                nj = d[factors[j] * gcds[d[x // factors[j]]][i]]
                if ndp[nj] > dp[j] + 1:
                    ndp[nj] = dp[j] + 1
                    flg = True
            
            if not flg:
                break
            
            dp = ndp
    
    print(dp[-1] if dp[-1] < inf else -1)

File: daily_problems/2026/08/0819/personal_submission/cf104544a_yawn_sean.cpp
#include <bits/stdc++.h>
// #include "atcoder/all"
// #pragma GCC optimize("Ofast")
// #pragma GCC optimize("unroll-loops")
// #pragma GCC target("sse,sse2,sse3,ssse3,sse4,popcnt,abm,mmx,avx2,tune=native")

using namespace std;

auto rng = mt19937(random_device()());
auto rngl = mt19937_64(random_device()());

int main() {
	ios_base::sync_with_stdio(false);
	cin.tie(0);
	cout.tie(0);

	int n, x;
	cin >> n >> x;

	vector<int> factors;

	for (int i = 1; i <= 100000; i ++) {
		if (x / i < i) break;
		if (x % i == 0) {
			factors.emplace_back(i);
			if (x / i != i) factors.emplace_back(x / i);
		}
	}

	int k = factors.size();
	sort(factors.begin(), factors.end());

	auto pos = [&] (int x) -> int {
		return lower_bound(factors.begin(), factors.end(), x) - factors.begin();
	};

	vector<int> cnt(k, 0);

	for (int i = 0; i < n; i ++) {
		long long v; cin >> v;
		cnt[pos(gcd(v, x))] ++;
	}

	vector<vector<int>> mapping(k, vector<int>(k));

	for (int i = 0; i < k; i ++) {
		for (int j = 0; j < k; j ++) {
			mapping[i][j] = pos(factors[i] * gcd(x / factors[i], factors[j]));
		}
	}

	vector<int> dp(k, n + 5);
	dp[0] = 0;

	for (int i = 0; i < k; i ++) {
		while (cnt[i] --) {
			vector<int> ndp = dp;
			bool flg = false;

			for (int j = 0; j < k; j ++) {
				int nj = mapping[i][j];
				if (ndp[nj] > dp[j] + 1) {
					ndp[nj] = dp[j] + 1;
					flg = true;
				}
			}

			if (!flg) break;

			dp.swap(ndp);
		}
	}

	cout << (dp[k - 1] <= n ? dp[k - 1] : -1);

	return 0;
}